task/task.go

```diff
@@ -0,0 +1,246 @@
+// Package task provides tools to manage running a process & capturing its
+// output.
+package task
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/tcsc/levity/api"
+)
+
+// ErrInvalidState indicates that an operation was attempted on a task
+// in a state not prepared for it
+var ErrInvalidState = errors.New("task in invalid state for operation")
+
+// streamReader catches the output from one of a Cmd's output streams (i.e.
+// stdout or stderr) and writes it out to a byte buffer in a Task, under
+// a write lock.
+type streamReader struct {
+	lock *sync.RWMutex
+	dst  *bytes.Buffer
+}
+
+func (r *streamReader) Write(b []byte) (n int, err error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.dst.Write(b)
+}
+
+// Task represents a task that has been invoked by the API server.
+type Task struct {
+	lock       sync.RWMutex
+	cmd        *exec.Cmd
+	stdout     bytes.Buffer
+	stderr     bytes.Buffer
+	statusCode api.TaskStatusCode
+	exitCode   int
+	done       chan struct{}
+}
+
+// New creates (but does not start) new task
+func New(binary string, workingDir string, env map[string]string, args ...string) *Task {
+	// Construct the underlying command which will do the heavy lifting of executing
+	// the subcommand.
+	cmd := exec.Command(binary, args...)
+	cmd.Dir = workingDir
+	cmd.Env = append(os.Environ(), formatEnvironment(env)...)
+
+	// wrap it in a Task to provide locking, and bind the output streams to readers
+	// that will capture the stream data and write it to the given buffers
+	t := Task{
+		cmd:        cmd,
+		statusCode: api.TaskStatusCode_NotStarted,
+		done:       make(chan struct{}),
+		exitCode:   -1,
+	}
+	t.cmd.Stdout = &streamReader{lock: &t.lock, dst: &t.stdout}
```

Comment:
Similar to question above, why not just use `bytes.Buffer` here?

Reply:
See above.